---
title: "RADseq data assembly in `ipyrad`"
knit: (function(input_file, encoding) {
  out_dir <- 'docs';
  rmarkdown::render(input_file,
 encoding=encoding,
 output_file=file.path(dirname(input_file), out_dir, 'index.html'))})
author: "Todd W. Pierson"
date: "16 October 2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, include=FALSE}
setwd("/Volumes/G-DRIVE/Dropbox/Teaching/2019_NRES_721/nres721_assembly")
```

<div class="alert alert-info">
  <strong>Note:</strong> This tutorial is written to be a basic introduction to assembling RADseq-style data in `ipyrad` for an in-class demonstration, not to be a comprehensive guide to doing this for research purposes. If you wish to follow along, please install the latest version of [`ipyrad`](https://ipyrad.readthedocs.io/); I have written this using `v0.7.30`. This program is written for Linux or Mac operating systems; if you are running a Windows machine, you can try to run `ipyrad` through your favorite emulator or virtual machine, or you can just follow along as we move through the tutorial. We'll return to `R` for downstream analyses!
</div>

In this tutorial, we'll practice assembling RADseq-style data. Our objectives are to:

1) understand the broad strokes of how `ipyrad` assembles data
2) compare results from a de novo and reference-based assembly
3) produce output files for use in downstream analyses

To do this, we'll use data from the patch-nosed salamander (*Urspelerpes brucei*). This is a tiny (~ 25 mm) lungless salamander endemic to just ~ 20 km^2^ in Georgia and South Carolina. Aren't they beautiful? 

<center>![](https://live.staticflickr.com/277/19325478625_f58f9b7b15_b.jpg){ width=40% }</center>

\
Today, we're going to use a small subset of a larger, unpublished RADcap [(Hoffberg et al. 2016)](https://onlinelibrary.wiley.com/doi/abs/10.1111/1755-0998.12566) dataset generated from individuals sampled across the small distribution of this species. This subset contains data from twelve individuals: three each from four collection localities. We can see these files in our `/data` directory:

```{bash, comment=NA}
ls ./data/*
```

There are a total of 24 files, consisting of a R1 and R2 file for each of the nine samples. Samples U03--U05 come from one collection locality, U19--U21 from a second, U23--U25 from a third, and U41--U43 from a fourth. These data have already been demultiplexed and "decloned" (i.e., PCR duplicates removed); the data for each sample has been subsetted down to 37,500 reads for consistently and manageability in this tutorial. 

There are two general ways in which we can assembled these data: de novo and using a reference genome. In this case, there does not exist a full reference genome for any close relative of *Urspelerpes*; because salamanders have very large genomes, few full assemblies exist. First, we'll try assembling these data de novo.

Let's create a new assembly by typing `ipyrad -n ubrucei_denovo`. Let's look at this template `ipyrad` parameters file:

```{bash, comment=NA}
cat ipyrad_files/params-ubrucei_denovo.txt
```

This is the file in which you can tweak a number of assembly parameters. See [here](https://ipyrad.readthedocs.io/parameters.html) for a thorough description of what each of these means. For our purposes, we only need to change a few. In your favorite text editor, open `params-ubrucei_denovo.txt` and change the following lines:

* Write `./data/*fq.gz` for `[2] [raw_fastq_path]`; this is telling `ipyrad` to use all .fq.gz files in our `data/` directory.
* `pairddrad` for `[7] [datatype]`; although these are RADcap data and thus, `pair3rad` might seem most appropriate, we have already independently "decloned" the reads and will treat them as `pairddrad` data for the purposes of this tutorial.
* `[8] [restriction_overhang]`
.....

Next, we'll assemble the data. [insert explanation]

```{bash, eval = FALSE}
ipyrad -p params-ubrucei_denovo.txt -s 1
```

[evaluate results]

Although we don't have a reference genome, we can assemble these data against a "pseudo-reference" genome that consists of the RADcap loci for which we designed baits.

```{bash, eval = FALSE, eval = FALSE}
cp params-ubrucei_denovo.txt params-ubrucei_reference.txt
```

```{bash, eval = FALSE}
ipyrad -p params-ubrucei_reference.txt -s 1234567
```